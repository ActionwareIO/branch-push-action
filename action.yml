name: 'Branch push action'
description: 'Move files to a different branch and make a commit. Useful for GitHub Action deployment.'
inputs:
  github_token:
    required: true
runs:
  using: 'node12'
  main: 'dist/index.js'
