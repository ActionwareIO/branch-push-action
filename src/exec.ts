import { exec } from '@actions/exec'

export type Exec = (cmd: string) => Promise<string>

export function getExec(cwd: string): Exec {
  return async (cmd: string) => {
    let output = ''

    await exec(cmd, [], {
      cwd,
      listeners: {
        stdout: (data: Buffer) => {
          output += data.toString()
        },
      },
    })

    return output
  }
}
