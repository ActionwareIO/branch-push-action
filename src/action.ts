import { copySync } from 'fs-extra'
import { join } from 'path'
import { dirSync as tmp } from 'tmp'

import { Exec } from './exec'
import { checkIfRemoteBranchExists, isStageAreEmpty, newPristineBranch, setupGit, switchBranch } from './git/utils'

interface ActionCtx {
  exec: Exec
  env: NodeJS.ProcessEnv
  cwd: string
}

interface Options {
  branchName: string
  files: string[]
}

export async function action(ctx: ActionCtx, options: Options) {
  const tmpDir = tmp().name
  console.log('Tmp dir created: ', tmpDir)

  console.info(`Coping ${options.files.length} files away to tmp dir`)
  for (const file of options.files) {
    const fullOutputPath = join(tmpDir, file)

    copySync(join(ctx.cwd, file), fullOutputPath)
  }

  await setupGit(ctx.exec, ctx.env)
  if (await checkIfRemoteBranchExists(ctx.exec, options.branchName)) {
    console.log(`Remote branch ${options.branchName} exists. Switching...`)
    await switchBranch(ctx.exec, options.branchName)
  } else {
    console.log(`Remote branch ${options.branchName} does not exist. Creating pristine branch ${options.branchName}.`)
    await newPristineBranch(ctx.exec, options.branchName)
  }

  console.info(`Coping ${options.files.length} files back to the workspace and git adding them`)
  for (const file of options.files) {
    const fullOutputPath = join(tmpDir, file)

    copySync(fullOutputPath, join(ctx.cwd, file), { overwrite: true })

    await ctx.exec(`git add --force ${file}`)
  }

  if (!(await isStageAreEmpty(ctx.exec))) {
    await ctx.exec(`git commit -m "Automated release" -a`)
    await ctx.exec(`git push origin-authorized`)
  }
}
