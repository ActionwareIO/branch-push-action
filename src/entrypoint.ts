import * as core from '@actions/core'

import { action } from './action'
import { getExec } from './exec'

export async function entrypoint() {
  const cwd = process.cwd()
  const env = process.env

  await action({ cwd, exec: getExec(process.cwd()), env }, { branchName: 'action', files: [] })
}

entrypoint().catch((e) => {
  console.log('Error happened:', e)
  // in case of error mark action as failed
  core.setFailed(e.message)
})
