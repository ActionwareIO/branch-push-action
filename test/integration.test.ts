import { expect } from 'earljs'
import { readFileSync } from 'fs-extra'
import { join } from 'path'

import { action } from '../src/action'
import { getFilteringExec, makeWorkspace, writeFiles } from './helpers'

describe('integration', () => {
  it('creates new deploy branch', async () => {
    const workspaceFiles = {
      'action.yml': `name: 'action for tests'`,
      'dist/index.js': `console.log('test!')`,
      '.gitignore': 'dist/index.js',
    }
    const { exec, git, workspacePath } = await makeWorkspace(workspaceFiles)
    const filteringExec = getFilteringExec(exec)

    await exec('git add .gitignore action.yml')
    await exec('git commit -m init')

    await action(
      { cwd: workspacePath, env: {}, exec: filteringExec },
      { branchName: 'action', files: ['action.yml', 'dist/index.js'] },
    )

    const distIndexContents = readFileSync(join(workspacePath, 'dist/index.js'), 'utf-8')
    expect(distIndexContents).toEqual(workspaceFiles['dist/index.js'])

    const status = await git.status()
    // .gitignore should be not tracked on this branch
    // @todo due to earl back this requires ...
    expect({ ...status }).toBeAnObjectWith({ created: [], deleted: [], modified: [], not_added: ['.gitignore'] })

    const branchesInfo = await git.branch()
    // @todo due to earl back this requires ...
    expect({ ...branchesInfo }).toBeAnObjectWith({ all: ['action', 'master'], current: 'action' })

    const exactOutput = await exec('git diff-tree --no-commit-id --name-status -r HEAD')
    expect(exactOutput).toMatchSnapshot()
  })

  it('pushes to already existing branch', async () => {
    const workspaceFiles = {
      'action.yml': `name: 'action for tests'`,
      '.gitignore': 'dist/index.js',
    }
    const { exec, git, workspacePath } = await makeWorkspace(workspaceFiles)
    const filteringExec = getFilteringExec(exec)

    await exec('git add .gitignore action.yml')
    await exec('git commit -m init')
    // create action branch, commit old code and switch back to master
    await exec('git checkout -b action')
    writeFiles({ 'dist/index.js': `console.log('old!')` }, workspacePath)
    await exec('git add -f dist/index.js')
    await exec('git commit -m "some old build"')
    await exec('git checkout master')
    writeFiles({ 'dist/index.js': `console.log('new!')` }, workspacePath)

    await action(
      { cwd: workspacePath, env: {}, exec: filteringExec },
      { branchName: 'action', files: ['action.yml', 'dist/index.js'] },
    )

    const distIndexContents = readFileSync(join(workspacePath, 'dist/index.js'), 'utf-8')
    expect(distIndexContents).toEqual(`console.log('new!')`)

    const status = await git.status()
    // .gitignore should be not tracked on this branch
    // @todo due to earl back this requires ...
    expect({ ...status }).toBeAnObjectWith({ created: [], deleted: [], modified: [], not_added: [] })

    const branchesInfo = await git.branch()
    // @todo due to earl back this requires ...
    expect({ ...branchesInfo }).toBeAnObjectWith({ all: ['action', 'master'], current: 'action' })

    const exactOutput = await exec('git diff-tree --no-commit-id --name-status -r HEAD')
    expect(exactOutput).toMatchSnapshot()
  })

  it('works when nothing to commit', async () => {
    const workspaceFiles = {
      'action.yml': `name: 'action for tests'`,
      '.gitignore': 'dist/index.js',
    }
    const { exec, git, workspacePath } = await makeWorkspace(workspaceFiles)
    const filteringExec = getFilteringExec(exec)

    await exec('git add .gitignore action.yml')
    await exec('git commit -m init')
    // create action branch, commit old code and switch back to master
    await exec('git checkout -b action')
    writeFiles({ 'dist/index.js': `console.log('new!')` }, workspacePath)
    await exec('git add -f dist/index.js')
    await exec('git commit -m "some old build"')
    await exec('git checkout master')
    writeFiles({ 'dist/index.js': `console.log('new!')` }, workspacePath)

    await action(
      { cwd: workspacePath, env: {}, exec: filteringExec },
      { branchName: 'action', files: ['action.yml', 'dist/index.js'] },
    )

    const distIndexContents = readFileSync(join(workspacePath, 'dist/index.js'), 'utf-8')
    expect(distIndexContents).toEqual(`console.log('new!')`)

    const status = await git.status()
    // .gitignore should be not tracked on this branch
    // @todo due to earl back this requires ...
    expect({ ...status }).toBeAnObjectWith({ created: [], deleted: [], modified: [], not_added: [] })

    const branchesInfo = await git.branch()
    // @todo due to earl back this requires ...
    expect({ ...branchesInfo }).toBeAnObjectWith({ all: ['action', 'master'], current: 'action' })

    const exactOutput = await exec('git diff-tree --no-commit-id --name-status -r HEAD')
    expect(exactOutput).toMatchSnapshot()
  })
})
