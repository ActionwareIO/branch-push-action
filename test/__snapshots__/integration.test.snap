// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration creates new deploy branch 1`] = `
"A	action.yml
A	dist/index.js
"
`;

exports[`integration pushes to already existing branch 1`] = `
"M	dist/index.js
"
`;

exports[`integration works when nothing to commit 1`] = `
"A	dist/index.js
"
`;
